File: src/StringProcessor.js
export default class Vector extends Float32Array {
    init(i = 0) {
        this.fill(i);
    }
    
    copy(vec2) {
        for (let i = 0; i < this.length; i++) {
            this[i] = vec2[i];
        }
    }
    
    add(vec2, outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = this[i] + vec2[i];
        }
    }
    
    minus(vec2, outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = this[i] - vec2[i];
        }
    }

    mul(vec2, outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = this[i] * vec2[i];
        }
    }
    
    sdiv(scalar, outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = this[i] / scalar;
        }
    }
    
    cube(outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = this[i] * this[i] * this[i];
        }
    }
    
    Dmin(h, outvec, scaling = 1){
        outvec[0] = this[0] / h * scaling;
        for (let i = 1; i < this.length; i++) {
            outvec[i] = (this[i] - this[i-1]) / h * scaling;
        }
        outvec[this.length] = -this[this.length-1] / h * scaling;
    }
    
    Dplus(h, outvec, scaling = 1){
        for (let i = 0; i < this.length-1; i++) {
            outvec[i] = (this[i+1] - this[i]) / h * scaling;
        }
    }
    
    dot(vec2, out) {
        out.value = 0;
        for (let i = 0; i < this.length; i++) {
            out.value += this[i] * vec2[i];
        }
    }
    
    norm1(out) {
        out.value = 0;
        for (let i = 0; i < this.length; i++) {
            out.value += Math.abs(this[i]);
        }
    }
    
    sign(outvec) {
        for (let i = 0; i < this.length; i++) {
            outvec[i] = Math.sign(this[i]);
        }
    }
}

class StringProcessor extends AudioWorkletProcessor {
    constructor() {
        super();
        // Default physical constants
        this.l0 = 1;
        this.T = 40;
        this.rho = 8000;
        this.E = 2e11;
        this.R = 2.9e-4;
        this.s_u0 = 0.5;
        this.s_u1 = 0;
        
        // Derived physical constants
        this.updateDerivedConstants();
        
        // Higher level parameters
        this.t60_0 = 12;
        this.t60_1000 = 6;
        this.setDissFromDecays();
        
        this.f0 = 50;
        this.beta = 0.0001;
        this.setTandLFromf0Beta();
        
        // Sav control setting
        this.alpha = 0.8;
        this.lambda0 = 100;
        
        // Excitation/Listening position
        this.posex = 0.5;
        this.poslist = 0.5;
        
        // State variables
        this.dspInit(sampleRate);
        
        console.log(this);
    }
    
    dspInit(samplerate) {
        // Recompute physical coefficients from high level parameters
        this.setDissFromDecays();
        this.setTandLFromf0Beta();
        this.updateDerivedConstants();
        
        // Stability condition
        this.sr = samplerate;
        this.dt = 1 / this.sr;
        this.h = Math.sqrt(
            (this.T * Math.pow(this.dt, 2) 
            + Math.sqrt(Math.pow(this.T, 2) * Math.pow(this.dt, 4) + 16 * this.mu * this.E * this.I * Math.pow(this.dt, 2))) 
            / (2*this.mu)
        );
        this.N = Math.floor(Math.max(3, Math.min(1000, this.alpha * Math.floor(this.l0 / this.h))));
        // this.N = 30;
        console.log("N = ",this.N);
        this.h = this.l0 / this.N;
        
        this.updateCoefficients();
        
        // Initialize vectors
        this.qlast = new Vector(this.N-1);
        this.qmid = new Vector(this.N-1);
        this.qnow = new Vector(this.N-1);
        this.qnext = new Vector(this.N-1);
        this.g = new Vector(this.N-1);
        this.dxq =new Vector(this.N);
        this.dxq3 = new Vector(this.N);
        this.Vprime = new Vector(this.N-1);
        this.righthand = new Vector(this.N-1);
        this.dtq = new Vector(this.N-1);
        
        this.qlast.init();
        this.qmid.init();
        this.qnow.init();
        this.qnext.init();
        this.g.init();
        this.dxq.init();
        this.dxq3.init();
        this.Vprime.init();
        this.righthand.init();
        this.dtq.init();
        
        this.psi = 0;
        this.epsilon = 0;
        this.basegmod= 0;
        this.scaleshermann = 0;

        this.gdotqlast = {value: 0};
        this.gdotg = {value: 0};
        this.gdotrighthand = {value: 0};
        this.temp = {value: 0};

    }
    
    process(inputs, outputs) {
        // By default, the node has single input and output.
        const input = inputs[0];
        const output = outputs[0];
        if (input.length > 0) {
            for (let i = 0; i < output[0].length; i++) {
                output[0][i] = this.processSample(input[0][i]);
            }
            
            for (let channel = 1; channel < output.length; ++channel) {
                output[channel].set(output[0]);
            }
        } else {
            for (let i = 0; i < output[0].length; i++) {
                output[0][i] = this.processSample(0);
            }
            
            for (let channel = 1; channel < output.length; ++channel) {
                output[channel].set(output[0]);
            }
        }
        return true;
    }
    
    processSample(input) {
        // // SAV term (20% CPU)
        this.qnow.Dmin(this.h, this.dxq);
        
        this.dxq.cube(this.dxq3);
        
        const EA_T = this.E * this.A - this.T;
        this.dxq3.Dplus(this.h, this.Vprime, -(EA_T) / 2 * this.h);
        //
        this.V = 0;
        for (let i = 0; i < this.N; i++) {
            this.V += this.dxq3[i] * this.dxq[i] * this.h * EA_T / 8;
        }
        this.Vprime.sdiv(Math.sqrt(2 * this.V) + 1e-12, this.g);
        
        // Additional control term
        this.qnow.add(this.qlast, this.qmid);
        this.qmid.Dmin(this.h, this.dxq, 0.5);
        this.dxq.cube(this.dxq3);
        
        this.V = 0;
        for (let i = 0; i < this.N; i++) {
            this.V += this.dxq3[i] * this.dxq[i] * this.h * EA_T / 8;
        }
        
        this.epsilon = this.psi - Math.sqrt(2 * this.V);
        this.qnow.minus(this.qlast, this.dtq);
        this.dtq.norm1(this.temp);
        this.basegmod = - this.epsilon * this.lambda0 * this.dt / (this.temp.value + 1e-12);
        
        for (let i = 0; i < this.N-1; i++) {
            this.g[i] = this.g[i] + this.basegmod * Math.sign(this.dtq[i]);
        }
        
        
        // Filling righthand
        this.g.dot(this.qlast, this.gdotqlast);
        this.g.dot(this.g, this.gdotg);
        
        
        for (let i = 2; i < this.N-3; i++) {
            this.righthand[i] = this.current0 * this.qnow[i] 
            + this.last0 * this.qlast[i] 
            + this.current1 * (this.qnow[i+1] + this.qnow[i-1]) 
            + this.current2 * (this.qnow[i+2] + this.qnow[i-2]) 
            + this.last1 * (this.qlast[i+1] + this.qlast[i-1])
            + this.dt2/4 * 1 / this.h * this.g[i] * this.gdotqlast.value
            - this.dt2 * 1 / this.h * this.g[i] * this.psi;
        }
        // Boundaries
        this.righthand[0] = this.current0Bounds * this.qnow[0] 
            + this.last0 * this.qlast[0] 
            + this.current1 * (this.qnow[1]) 
            + this.current2 * (this.qnow[2]) 
            + this.last1 * (this.qlast[1])
            + this.dt2/4 * 1 / this.h * this.g[0] * this.gdotqlast.value
            - this.dt2 * 1 / this.h * this.g[0] * this.psi;
        
        this.righthand[this.N-2] = this.current0Bounds * this.qnow[this.N-2] 
            + this.last0 * this.qlast[this.N-2] 
            + this.current1 * (this.qnow[this.N-3]) 
            + this.current2 * (this.qnow[this.N-4]) 
            + this.last1 * (this.qlast[this.N-3])
            + this.dt2/4 * 1 / this.h * this.g[this.N-2] * this.gdotqlast.value
            - this.dt2 * 1 / this.h * this.g[this.N-2] * this.psi;
        
        this.righthand[1] = this.current0 * this.qnow[1]
            + this.last0 * this.qlast[1] 
            + this.current1 * (this.qnow[2] + this.qnow[0]) 
            + this.current2 * (this.qnow[3]) 
            + this.last1 * (this.qlast[2] + this.qlast[0])
            + this.dt2/4 * 1 / this.h * this.g[1] * this.gdotqlast.value
            - this.dt2 * 1 / this.h * this.g[1] * this.psi;
        
        this.righthand[this.N-3] = this.current0 * this.qnow[this.N-3]
            + this.last0 * this.qlast[this.N-3] 
            + this.current1 * (this.qnow[this.N-4] + this.qnow[this.N-2]) 
            + this.current2 * (this.qnow[this.N-5]) 
            + this.last1 * (this.qlast[this.N-4] + this.qlast[this.N-2])
            + this.dt2/4 * 1 / this.h * this.g[this.N-3] * this.gdotqlast.value
            - this.dt2 * 1 / this.h * this.g[this.N-3] * this.psi;
        
        
        // Excitation 
        this.righthand[Math.floor(this.posex * (this.N-2))] += this.dt2 * input / this.h;
        
        // Solving using shermann morrison
        this.g.dot(this.righthand, this.gdotrighthand);
        this.scaleshermann = - this.dt2/4 * 1 / this.h * this.gdotrighthand.value / (1 + this.term0 * this.dt2/4 * 1 / this.h * this.gdotg.value); 
        
        for (let i = 0; i < this.N-1; i++) {
            this.qnext[i] = this.term0 * (this.righthand[i] + this.term0 * this.scaleshermann* this.g[i]);
        }
        
        this.qnext.minus(this.qlast, this.dtq);
        this.g.dot(this.dtq, this.temp);

        this.psi = this.psi + 0.5 * this.temp.value;

        
        
        // Update state variables
        
        this.qlast.set(this.qnow);
        this.qnow.set(this.qnext);
        return 0.1*(this.qnow[Math.floor(this.poslist * (this.N-2))] - this.qlast[Math.floor(this.poslist * (this.N-2))]) / (this.dt * Math.sqrt(this.T * this.mu));
    }
    
    setDissFromDecays(){
        const gamma2 = this.T / this.mu;
        const kappa2 = this.E * this.I / (this.mu);
        const beta2_0 = this.zeta(2 * Math.PI, gamma2, kappa2);
        const beta2_1 = this.zeta(2 * Math.PI * 1000, gamma2, kappa2);
        
        this.s_u0 = Math.max(0, 6 * Math.log(10) / (beta2_1 - beta2_0) * (beta2_1 / this.t60_0 - beta2_0 / this.t60_1000));
        this.s_u1 = Math.max(0, 6 * Math.log(10) / (beta2_1 - beta2_0) * (-1 / this.t60_0 + 1 / this.t60_1000));
        console.log("s_u0 = ", this.s_u0, "s_u1 = ", this.s_u1);
    }
    
    zeta(omega, gamma2, kappa2){
        return (-gamma2 + Math.sqrt(gamma2 * gamma2 + 4 * omega * omega * kappa2)) / (2 * kappa2);
    }
    
    setTandLFromf0Beta(){
        this.T = Math.sqrt(
            (Math.pow(Math.PI, 4) * this.E * this.f0 * this.f0 * Math.pow(this.R, 6) * this.rho) / (this.beta * (1 + this.beta))
        );
        this.l0 = Math.sqrt(
            Math.pow(Math.PI, 3) * this.E * Math.pow(this.R, 4) / (4 * this.beta * this.T)
        );
    }
    
    updateDerivedConstants(){
        this.A = Math.PI * this.R * this.R;
        this.I = Math.PI * this.R * this.R * this.R * this.R / 4;
        this.mu = this.rho * this.A;
    }
    
    updateCoefficients(){
        this.current0 = 2 * this.mu 
        - 2 * Math.pow(this.dt/this.h, 2) * this.T
        - 4 * this.mu * this.s_u1 * this.dt / Math.pow(this.h, 2) 
        - Math.pow(this.dt, 2) * this.E * this.I * 6/Math.pow(this.h, 4);
        this.current0Bounds = this.current0 + Math.pow(this.dt, 2) * this.E * this.I / Math.pow(this.h, 4);
        
        this.last0 = -(1 - this.dt * this.s_u0) * this.mu
        + 4 * this.mu * this.dt * this.s_u1 / Math.pow(this.h, 2);
        this.current1 = Math.pow(this.dt / this.h, 2) * this.T 
        + Math.pow(this.dt, 2) * this.E * this.I * 4 / Math.pow(this.h, 4)
        + 2 * this.mu * this.s_u1 * this.dt / Math.pow(this.h, 2);
        this.current2 = - Math.pow(this.dt, 2) * this.E * this.I / Math.pow(this.h, 4);
        this.last1 = - 2 * this.mu * this.s_u1 * this.dt / (Math.pow(this.h, 2));
        
        this.term0 = 1 / (this.mu * (1 + this.dt * this.s_u0));
        this.dt2 = Math.pow(this.dt, 2);
    }
}


registerProcessor('string-processor', StringProcessor);

File: src/processor/StringProcessor.cpp
#include "StringProcessor.h"
#include "vecMath.h"
#include "iostream"

template <class T>
StringProcessor<T>::StringProcessor(float sampleRate, bool controlTerm){
    this->controlTerm = controlTerm;
    // Default physical constants
    l0 = 1;
    T0 = 40;
    rho = 8000;
    E = 2e11;
    R = 2.9e-4;
    eta_0 = 0.5;
    eta_1 = 0;
        
    // Derived physical constants
    updateDerivedConstants();
        
    // Higher level parameters
    t60_0 = 12;
    t60_1 = 6;
    fd0 = 0;
    fd1 = 1000;
    setDissFromDecays();
        
    f0 = 100;
    beta = 1e-4;
    setTandLFromf0Beta();

    // Sav control setting
    alpha = 0.9;
    lambda0 = 100;

    // Bow curve parameters
    alphaBow = 10;
        
    // Excitation/Listening position
    posex = 0.72;
    poslistL = 0.3;
    poslistR = 0.3;

    // State variables
    reinitDsp(sampleRate);
};

template <class T>
void StringProcessor<T>::updateDerivedConstants() {
    A = M_PI * R * R;
    I = M_PI * R * R * R * R / 4;
    mu = rho * A;
}

template <class T>
void StringProcessor<T>::setTandLFromf0Beta() {
    T0 = sqrt( (pow(M_PI, 4)*E*pow(f0, 2)*pow(R, 6)*rho ) / (beta * (1+beta)) );
    l0 = sqrt( (pow(M_PI, 3)*E*pow(R, 4)) / (4 * beta * T0) );
}

template <class T>
void StringProcessor<T>::setDissFromDecays() {
    double gamma2 = T0/mu;
    double kappa2 = E*I/mu;
    double beta2_0 = zeta(2*M_PI*fd0, gamma2, kappa2);
    double beta2_1 = zeta(2*M_PI*fd1, gamma2, kappa2);

    eta_0 = fmax(3 * log(10) / (beta2_1 - beta2_0) * (beta2_1/t60_0 - beta2_0/t60_1), 0);
    eta_1 = fmax(3 * log(10) / (beta2_1 - beta2_0) * (-1/t60_0 + 1/t60_1), 0);
}

template <class T>
T StringProcessor<T>::zeta(T omega, T gamma2, T kappa2) {
    return (-gamma2 + sqrt(gamma2*gamma2 + 4*kappa2*omega*omega))/(2*kappa2);
}

template <class T>
void StringProcessor<T>::modifyhFromBend() {
    // Only the length is modified for simplicity.
    // Inharmonicity is not taken into account.
    fbend = f0 * pow(2, bend/1200);
    //T lbend = 1/(2*fbend) * sqrt(T0/mu) * sqrt(1 + pow(M_PI, 3) * E * pow(R, 4) / (4 * l0 * l0));
    T a = fbend * fbend;
    T b = - T0 / (4*mu);
    T c = - pow(M_PI, 2) * E * I / mu;
    T delta = pow(b, 2) - 4 * a * c;
    T lbend = sqrt((-b + sqrt(delta)) / (2 * a));
    T gamma = dt * dt * T0 + 4 * dt * mu * eta_1;
    h = fmax(lbend / N, sqrt(
        (gamma
        + sqrt(gamma*gamma + 16 * mu * E * I * pow(dt, 2))) 
        / (2*mu)
    ));
}

template <class T>
void StringProcessor<T>::reinitDsp(float sampleRate) {
    fbend = f0;
    // Recompute physical coefficients from high level parameters
    setDissFromDecays();
    setTandLFromf0Beta();
    updateDerivedConstants();

    modifyhFromBend();
    
    // Stability condition
    sr = sampleRate;
    dt = 1 / sr;
    T gamma = dt * dt * T0 + 4 * dt * mu * eta_1;
    h = sqrt(
        (gamma
        + sqrt(gamma*gamma + 16 * mu * E * I * pow(dt, 2))) 
        / (2*mu)
    );
    N = static_cast<int>(floor(fmax(3, fmin(1000, alpha * floor(l0 / h)))));
    h = l0 / N;
    
    updateCoefficients();
    
    // Initialisation of vectors to right size
    qlast = Eigen::Vector<T, -1>::Zero(N-1);
    qnow = Eigen::Vector<T, -1>::Zero(N-1);
    qnext = Eigen::Vector<T, -1>::Zero(N-1);
    g = Eigen::Vector<T, -1>::Zero(N-1);
            
    dxq = Eigen::Vector<T, -1>::Zero(N);
    dxq3 = Eigen::Vector<T, -1>::Zero(N);

    Vprime = Eigen::Vector<T, -1>::Zero(N-1);

    righthand = Eigen::Vector<T, -1>::Zero(N-1);

    Rbow = Eigen::Vector<T, -1>::Zero(N-1);
    
    psi = 0;
}
template <class T>
void StringProcessor<T>::updateCoefficients(){
        D40 = Eigen::Vector<T, -1>::Ones(N-1) * 6 / pow(h, 4);
        D40[0] = 5/ pow(h, 4);
        D40[N-2] = 5/ pow(h, 4);
        D41 = - 4/(pow(h, 4));
        D42 = 1/(pow(h, 4));
        Current0 = (2 * mu - 2 * pow(dt/h, 2) * T0
                    - 4 * mu * eta_1 * dt / pow(h, 2)) * Eigen::Vector<T, -1>::Ones(N-1)
                    - pow(dt, 2) * E * I * D40;
        Last0 = -(1 - dt * eta_0) * mu + 4 * mu * dt * eta_1 / pow(h, 2);
        Current1 = pow(dt/h, 2)*T0-pow(dt, 2) * E * I * D41 + 2*mu*eta_1*dt/pow(h, 2);
        Current2 = - pow(dt, 2) * E * I * D42;
        Last1 = - 2*mu*eta_1*dt/pow(h, 2);
}

template <class T>
void StringProcessor<T>::computeVAndVprime(){
    switch (nl_mode)
    {
        case 0:
            V = 0;
            Vprime.setZero();
            break;
        case 1:
            dxq.setZero();
            dxq.head(N-1) = qnow;
            dxq.tail(N-1) -= qnow;
            dxq /= h;
            dxq2 = dxq.dot(dxq);
            
            V = E * A / (8 * l0) * h * h * pow(dxq2, 2);
            Vprime = - E * A / (2 * l0) * h * dxq2 * (dxq.tail(N-1) - dxq.head(N-1));
            break;
        case 2:
            dxq.setZero();
            dxq.head(N-1) = qnow;
            dxq.tail(N-1) -= qnow;
            dxq /= h;
            dxq3 = dxq.array().cube();

            Vprime = -(E * A - T0) / 2 * (dxq3.tail(N-1) - dxq3.head(N-1));
            V = (E * A - T0) / 8 * h * (dxq3.cwiseProduct(dxq)).sum();
            break;
        case 3:
            V = 0;
            Vprime.setZero();
            break;
        default:
            V = 0;
            Vprime.setZero();
            break;
    }
}

template <class T>
void StringProcessor<T>::computeV(){
    switch (nl_mode)
    {
        case 0:
            V = 0;
            break;
        case 1:
            dxq.setZero();
            dxq.head(N-1) = (qnow + qlast) / 2;
            dxq.tail(N-1) -= (qnow + qlast) / 2;
            dxq /= h;

            V = E * A / (8 * l0) * h * h * pow(dxq.dot(dxq), 2);
            break;
        case 2:
            dxq.setZero();
            dxq.head(N-1) = (qnow + qlast) / 2;
            dxq.tail(N-1) -= (qnow + qlast) / 2;
            dxq /= h;
            dxq3 = dxq.array().cube();

            V = (E * A - T0) / 8 * h * (dxq3.cwiseProduct(dxq)).sum();
            break;
        case 3:
            V = 0;
            break;
        default:
            V = 0;
            break;
    }
}

template <class T>
std::tuple<T, T, T> StringProcessor<T>::process(T input, T bend, T posex, T poslistL, T poslistR, T t60_0) {
    //Eigen::internal::set_is_malloc_allowed(false);
    // Pitch bend
    if (bend != this->bend) {
        this->bend = bend;
        modifyhFromBend();
        updateCoefficients();
    }
    // Excitation and listening positions
    if (posex != this->posex || poslistL != this->poslistL || poslistR != this->poslistR) {
        this->posex = std::clamp(posex, T(0), T(1));
        this->poslistL = std::clamp(poslistL, T(0), T(1));
        this->poslistR = std::clamp(poslistR, T(0), T(1));
    }
    // Modify damping from t60_0. Only eta_0 is modified to preserve stability.
    // F60_0 is considered to be zero here.
    if (t60_0 != this->t60_0 && t60_0 > 1e-4) {
        this->t60_0 = t60_0;
        this->eta_0 = 3 * log(10) / (t60_0);

        updateCoefficients();
    }

    // Compute g (dependant on the nonlinear mode => value of nl)
    computeVAndVprime();
    g = Vprime / (sqrt(2 * V) + 1e-12);

    if (controlTerm) {
        computeV();
        epsilon = psi - sqrt(2 * V);
        g+= -lambda0 * epsilon *dt * ((qnow-qlast).array()>0).select(Eigen::Vector<T, -1>::Ones(N-1), -Eigen::Vector<T, -1>::Ones(N-1)) / ((qnow-qlast).template lpNorm<1>() + 1e-12);
    }

    // Linear part
    righthand = Current0.cwiseProduct(qnow) + Last0 * qlast;
    righthand.head(N-2) += Current1*qnow.tail(N-2) + Last1*qlast.tail(N-2);
    righthand.tail(N-2) += Current1*qnow.head(N-2) + Last1*qlast.head(N-2);

    righthand.tail(N-3) += Current2*qnow.head(N-3);
    righthand.head(N-3) += Current2*qnow.tail(N-3);
    
    // External force
    righthand(static_cast<int>(floor(this->posex*(N-2)))) += pow(dt, 2) * input / h * (1 - (this->posex*(N-2) - floor(this->posex*(N-2))));
    righthand(static_cast<int>(ceil(this->posex*(N-2)))) += pow(dt, 2) * input / h * ((this->posex*(N-2) - floor(this->posex*(N-2))));
    
    // Nonlinear part
    righthand +=  pow(dt/2, 2) * 1/h * g * g.dot(qlast) 
                -  pow(dt, 2) * 1/h * g * psi;

    // Solving using shermann morrisson
    double term0 = 1 / (mu*(1+dt*eta_0));
    qnext = term0 * righthand
        - pow(dt/2, 2)*pow(term0, 2) * 1/h * g * g.dot(righthand) / (1 + term0 *pow(dt/2, 2) * 1/h * g.dot(g));

    psi = psi + 0.5 * g.dot(qnext - qlast);

    qlast = qnow;
    qnow = qnext;
    // Eigen::internal::set_is_malloc_allowed(true);
    vout();
    return { vl, vr,  epsilon / (sqrt(2*V) + 1e-12) };
}

template <class T>
std::tuple<T, T, T> StringProcessor<T>::processBowed(T vbow, T Fbow, T bend, T posex, T poslistL, T poslistR) {
    vbow *= 1e-5;
    Fbow *= 1e-5;
    
    // Compute relative bow velocity
    vrel = (qnow(static_cast<int>(posex * (N-2)))- qlast(static_cast<int>(poslistL * (N-2))))/dt-vbow;
    // Compute bow force
    phinow = phi(vrel);
    Rbow.setZero();
    Rbow(static_cast<int>(posex * (N-2))) = abs(Fbow) * phinow / (vrel + std::copysign(1e-12, vrel));

    //Eigen::internal::set_is_malloc_allowed(false);
    // Pitch bend
    if (bend != this->bend) {
        this->bend = bend;
        modifyhFromBend();
        updateCoefficients();
    }
    // Excitation and listening positions
    if (posex != this->posex || poslistL != this->poslistL || poslistR != this->poslistR) {
        this->posex = std::clamp(posex, T(0), T(1));
        this->poslistL = std::clamp(poslistL, T(0), T(1));
        this->poslistR = std::clamp(poslistR, T(0), T(1));
    }

    // Compute g
    dxq.setZero();
    dxq.head(N-1) = qnow;
    dxq.tail(N-1) -= qnow;
    dxq /= h;

    dxq3 = dxq.array().cube();

    Vprime = -(E * A - T0) / 2 * (dxq3.tail(N-1) - dxq3.head(N-1));
    V = (E * A - T0) / 8 * h * (dxq3.cwiseProduct(dxq)).sum();

    //g = Vprime / (sqrt(2 * V) + 1e-12);
    
    // G modification with regularisation term
    if (controlTerm) {
        dxq.setZero();
        dxq.head(N-1) = (qnow + qlast) / 2;
        dxq.tail(N-1) -= (qnow + qlast) / 2;
        dxq /= h;
        dxq3 = dxq.array().cube();
        V = (E * A - T0) / 8 * h * (dxq3.cwiseProduct(dxq)).sum();
        epsilon = psi - sqrt(2*V);
        g += -lambda0 * epsilon *dt * ((qnow-qlast).array()>0).select(Eigen::Vector<T, -1>::Ones(N-1), -Eigen::Vector<T, -1>::Ones(N-1)) / ((qnow-qlast).template lpNorm<1>() + 1e-12);
    }
    // Linear part
    righthand = Current0.cwiseProduct(qnow) + (Last0* Eigen::Vector<T, -1>::Ones(N-1) + dt* Rbow * h * 0.5).cwiseProduct(qlast) - Rbow * vrel;
    righthand.head(N-2) += Current1*qnow.tail(N-2) + Last1*qlast.tail(N-2);
    righthand.tail(N-2) += Current1*qnow.head(N-2) + Last1*qlast.head(N-2);

    righthand.tail(N-3) += Current2*qnow.head(N-3);
    righthand.head(N-3) += Current2*qnow.tail(N-3);
    
    // External force
    // righthand(static_cast<int>(floor(this->posex*(N-2)))) += pow(dt, 2) * input / h * (1 - (this->posex*(N-2) - floor(this->posex*(N-2))));
    // righthand(static_cast<int>(ceil(this->posex*(N-2)))) += pow(dt, 2) * input / h * ((this->posex*(N-2) - floor(this->posex*(N-2))));
    
    // Nonlinear part
    righthand +=  pow(dt/2, 2) * 1/h * g * g.dot(qlast)
                -  pow(dt, 2) * 1/h * g * psi;

    // Solving using shermann morrisson
    term0V = (mu*((1+dt*eta_0)* Eigen::Vector<T, -1>::Ones(N-1)+dt* Rbow * h * 0.5)).cwiseInverse();
    qnext = term0V.cwiseProduct(righthand)
        - pow(dt/2, 2)* 1/h *term0V.cwiseProduct(term0V).cwiseProduct(g) * g.dot(righthand) / (1 + pow(dt/2, 2) * 1/h * term0V.cwiseProduct(g).dot(g));

    psi = psi + 0.5 * g.dot(qnext - qlast);

    qlast = qnow;
    qnow = qnext;
    // Eigen::internal::set_is_malloc_allowed(true);
    vout();
    return { vl, vr,  epsilon / (sqrt(2*V) + 1e-12) };
}

template <class T>
T StringProcessor<T>::phi(T vrel) {
    return sqrt(2*alphaBow)*vrel*exp(-alphaBow*vrel*vrel + 0.5);
}

template <class T>
void StringProcessor<T>::vout() {
    vl = (
            (qnow(static_cast<int>(floor(poslistL * (N-2) )))- qlast(static_cast<int>(floor(poslistL * (N-2))))) * (1 - (poslistL*(N-2) - floor(poslistL*(N-2))))
            + (qnow(static_cast<int>(ceil(poslistL * (N-2) )))- qlast(static_cast<int>(ceil(poslistL * (N-2))))) * (poslistL*(N-2) - floor(poslistL*(N-2)))
         ) / (dt * sqrt(T0*mu));
    vr = (
            (qnow(static_cast<int>(floor(poslistR * (N-2) )))- qlast(static_cast<int>(floor(poslistR * (N-2))))) * (1 - (poslistR*(N-2) - floor(poslistR*(N-2))))
            + (qnow(static_cast<int>(ceil(poslistR * (N-2) )))- qlast(static_cast<int>(ceil(poslistR * (N-2))))) * (poslistR*(N-2) - floor(poslistR*(N-2)))
         ) / (dt * sqrt(T0*mu));
}





template class StringProcessor<double>;
template class StringProcessor<float>;
